### PR TITLE
Add function to compute cross_entropy for 2D image

### PR DESCRIPTION
**Pull request by wkentaro**
This is used to compute loss for class segmentation task, for example.
Ignoring negative labels is required to use dataset with unlabeled region.

### CODE REVIEW

**Comment by apaszke** [comment]
Some time ago we've decided that we want to slowly start moving away from using Nd suffixes for loss and activation functions. Can you add that as a special condition to the `cross_entropy` function? Just use an `if` with an `input.dim()` condition.

File: torch/nn/functional.py [hunk 1]
```diff
@@ -439,16 +439,37 @@ def cross_entropy(input, target, weight=None, size_average=True):
...
-    return nll_loss(log_softmax(input), target, weight, size_average)
+    ndim = input.ndim()
+    if ndim == 2:
+        return nll_loss(log_softmax(input), target, weight, size_average)
+    elif ndim() == 4:
+        N, C, H, W = input.size()
+        log_p = F.log_softmax(input)  # (N, C, H, W)
+        log_p = log_p.transpose(1, 2).transpose(2, 3).contiguous().view(-1, C)
+        log_p = log_p[target.view(N, H, W, 1).repeat(1, 1, 1, C) >= 0]
+        log_p = log_p.view(-1, C)  # (M, C) where M = (target >= 0).sum()
```
Comment (wkentaro) [inline torch/nn/functional.py]: To support Nd tensor, the code will be something like below,
but currently not sure how to handle `...` with PyTorch.
Should I create size variable like `size = target.size(); size.insert(0, 1);`?

```
N, C = input.size()[:2]
log_p = F.log_softmax(input)  # (N, C, ...)
log_p = log_p.transpose(0, 1).contiguous().view(C, -1)  # (C, X)
log_p = log_p[target.view(1, ...).repeat(C, 1, 1, ..., 1) >= 0]  # (M*C,)
log_p = log_p.view(-1, C)  # (M, C) where M = (target >= 0).sum()
```

**Comment by wkentaro** [comment]
Thank you for the comment. I updated the commit.

File: torch/nn/functional.py (hunk 1 above)
Comment (apaszke) [inline torch/nn/functional.py]: I think you could use `target.unsqueeze(0).expand_as(log_p)`

Comment (apaszke) [inline torch/nn/functional.py]: BTW, we have NLLLoss2d op and it's going to be much faster to use that just like in the 2D case. But an Nd implementation as a fallback is welcome too!

File: torch/nn/functional.py [hunk 2]
```diff
@@ -439,16 +439,37 @@ def cross_entropy(input, target, weight=None, size_average=True):
...
+    dim = input.dim()
+    if dim == 2:
+        return nll_loss(log_softmax(input), target, weight, size_average)
+    elif dim == 4:
+        N, C, H, W = input.size()
+        log_p = F.log_softmax(input)  # (N, C, H, W)
+        log_p = log_p.transpose(1, 2).transpose(2, 3).contiguous().view(-1, C)
+        log_p = log_p[target.view(N, H, W, 1).repeat(1, 1, 1, C) >= 0]
+        log_p = log_p.view(-1, C)  # (M, C) where M = (target >= 0).sum()
+        mask = target >= 0  # (M,)
```
Comment (apaszke) [inline torch/nn/functional.py]: I don't think we support negative indices in targets in the 2D case

File: torch/nn/functional.py [hunk 3]
```diff
@@ -439,16 +439,37 @@ def cross_entropy(input, target, weight=None, size_average=True):
...
+        target = target[mask]
+        loss = F.nll_loss(log_p, target, weight=weight, size_average=False)
+        if size_average:
+            loss /= mask.sum().data[0]
+        else:
+            loss /= N
```
Comment (apaszke) [inline torch/nn/functional.py]: You shouldn't average the loss in any way if `size_average` is False

File: torch/nn/functional.py (hunk 1 above)
Comment (apaszke) [inline torch/nn/functional.py]: On second thought the (C, X) data format will be very very inefficient, so maybe let's just stick to the native kernels for now.

File: torch/nn/functional.py (hunk 2 above)
Comment (wkentaro) [inline torch/nn/functional.py]: I'd like to add a feature of ignoring some label values.
Is that possible without this?

File: torch/nn/functional.py (hunk 1 above)
Comment (wkentaro) [inline torch/nn/functional.py]: Thank you, I will try that.

File: torch/nn/functional.py (hunk 3 above)
Comment (wkentaro) [inline torch/nn/functional.py]: Ok.

File: torch/nn/functional.py (hunk 2 above)
Comment (apaszke) [inline torch/nn/functional.py]: I think it's going to be easier once we add element-wise losses. Or we could modify our C and CUDA backends. @soumith wdyt?

**Comment by wkentaro** [comment]
The `total_weight` argument seems the rescaling factor in element-wise.
Is this correct?

**Comment by apaszke** [comment]
I don't think it's always the case and I'd rather not expose this buffer. Can you remove the additional arg?

**Comment by wkentaro** [comment]
Ok, I removed it.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks!

**Comment by wkentaro** [comment]
@apaszke BTW, how can we apply masking in computation of cross_entropy? For example, I'd like to ignore `-1` in the `target`.

**Comment by apaszke** [comment]
You can shift all targets one index forward and set the weight of the first class to 0.

**Comment by fmassa** [comment]
@apaszke This is a solution, but maybe not ideal as it forces the user to have an extra class for classification (the ignored class), and then at test time it needs to narrow all but the selected class before applying softmax (or we could end up having ignored classes as our prediction).
It's possible, but maybe not super convenient.

**Comment by wkentaro** [comment]
Are there any examples of using gpu kernel?
I think I can try implementing that functionality with it.

**Comment by fmassa** [comment]
The GPU kernels are implemented in https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/SpatialClassNLLCriterion.cu and the C function calling it in https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/generic/SpatialClassNLLCriterion.cu
But the same changes would need to be done in the corresponding CPU implementation, and then the python interface would need to be changed. But before doing it, we should see if it's something we would want to add to core or not (because the weighting can mimic the masking, but it's not super convenient for a few reasons mentioned above)

**Comment by wkentaro** [comment]
Thanks!
I see. Currently, I'm using advanced indexing of boolean array to extract indices of interest (0 or positive label values in my case) from image, which returns 1d array, and then computing cross_entropy loss:
https://github.com/wkentaro/pytorch-fcn/blob/c5fe29ac8235befd186ef608fa4c6f5c51f56269/torchfcn/trainer.py#L25-L27
But I'm not sure if masking in kernel is better than that. Maybe it is faster than the boolean array indexing?
